token-metadata/example/tests/program_test.rs

```diff
@@ -0,0 +1,167 @@
+    let payer = Arc::new(keypair_clone(&context.payer));
+    let context = Arc::new(Mutex::new(context));
+
+    let client: Arc<dyn ProgramClient<ProgramBanksClientProcessTransaction>> =
+        Arc::new(ProgramBanksClient::new_from_context(
+            Arc::clone(&context),
+            ProgramBanksClientProcessTransaction,
+        ));
+    (context, client, payer)
+}
+
+pub async fn setup_mint<T: SendTransaction>(
+    program_id: &Pubkey,
+    mint_authority: &Pubkey,
+    decimals: u8,
+    payer: Arc<Keypair>,
+    client: Arc<dyn ProgramClient<T>>,
+) -> Token<T> {
+    let mint_account = Keypair::new();
+    let token = Token::new(
+        client,
+        program_id,
+        &mint_account.pubkey(),
+        Some(decimals),
+        payer,
+    );
+    token
+        .create_mint(mint_authority, None, vec![], &[&mint_account])
+        .await
+        .unwrap();
+    token
+}
+
+pub async fn setup_metadata(
+    context: &mut ProgramTestContext,
+    metadata_program_id: &Pubkey,
+    mint: &Pubkey,
+    token_metadata: &TokenMetadata,
+    metadata_keypair: &Keypair,
+    mint_authority: &Keypair,
+) {
+    let rent = context.banks_client.get_rent().await.unwrap();
+    let space = token_metadata.tlv_size_of().unwrap();
+    let rent_lamports = rent.minimum_balance(space);
+    let transaction = Transaction::new_signed_with_payer(
+        &[
+            system_instruction::create_account(
+                &context.payer.pubkey(),
+                &metadata_keypair.pubkey(),
+                rent_lamports,
+                space.try_into().unwrap(),
+                metadata_program_id,
+            ),
+            initialize(
+                metadata_program_id,
+                &metadata_keypair.pubkey(),
+                &Option::<Pubkey>::from(token_metadata.update_authority.clone()).unwrap(),
+                mint,
+                &mint_authority.pubkey(),
+                token_metadata.name.clone(),
+                token_metadata.symbol.clone(),
+                token_metadata.uri.clone(),
+            ),
+        ],
+        Some(&context.payer.pubkey()),
+        &[&context.payer, metadata_keypair, mint_authority],
+        context.last_blockhash,
+    );
+
+    context
+        .banks_client
+        .process_transaction(transaction)
+        .await
+        .unwrap();
+}
+
+#[allow(dead_code)]
+pub async fn setup_update_field(
+    context: &mut ProgramTestContext,
+    metadata_program_id: &Pubkey,
+    token_metadata: &mut TokenMetadata,
+    metadata: &Pubkey,
+    update_authority: &Keypair,
+    field: Field,
+    value: String,
+) {
+    let rent = context.banks_client.get_rent().await.unwrap();
+    let old_space = token_metadata.tlv_size_of().unwrap();
+    let old_rent_lamports = rent.minimum_balance(old_space);
+
+    token_metadata.update(field.clone(), value.clone());
+
+    let new_space = token_metadata.tlv_size_of().unwrap();
+    let new_rent_lamports = rent.minimum_balance(new_space);
```

Comment:
Ahhhh, there it is!

I think this is less intuitive to do on client-side, but willing to hear thoughts.

I've typically handled realloc additional rent within the program itself, which I guess someone could still do, but curious if it's something we want to include in the example program

Reply:
Yep, that's fair. Interfaces forcing me to abandon habits.

> A program should never require a payer and system program

This is good to know